File: demo/belief_update_demo.cpp
#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstddef>
#include <cstdlib>
#include <iomanip>
#include <iostream>
#include <optional>
#include <stdexcept>
#include <string>
#include <string_view>
#include <vector>

#include "belief_update/analysis/local_sensitivity.hpp"
#include "belief_update/analysis/posterior_comparison.hpp"
#include "belief_update/analysis/threshold_analysis.hpp"
#include "belief_update/core/bayes.hpp"
#include "belief_update/replay/replay.hpp"
#include "belief_update/stress/stress.hpp"
#include "belief_update/tree.hpp"
#include "belief_update/types.hpp"
#include "belief_update/update/update.hpp"

namespace {

using Clock = std::chrono::steady_clock;

struct BenchmarkResult {
  std::string_view name;
  double nanoseconds_per_update;
  double posterior;
};

[[nodiscard]] belief_update::LikelihoodInput make_input(
    double p_e_given_h, double p_e_given_not_h,
    belief_update::SourceType source_type, std::string source_id,
    belief_update::OverrideStatus override_status) {
  using namespace belief_update;
  return LikelihoodInput{
      LikelihoodValues{p_e_given_h, p_e_given_not_h}, std::nullopt,
      Provenance{source_type, SourceId{std::move(source_id)},
                 SourceVersion{"demo-v1"}, ProvenanceTimestamp{1'000}, 0.95,
                 override_status}};
}

[[nodiscard]] belief_update::Provenance make_tree_provenance() {
  using namespace belief_update;
  return Provenance{SourceType::model, SourceId{"transport-model"},
                    SourceVersion{"demo-v1"}, ProvenanceTimestamp{1'000}, 0.95,
                    OverrideStatus::not_requested};
}

[[nodiscard]] double unchecked_formula(double prior, double p_e_given_h,
                                       double p_e_given_not_h) noexcept {
  const double numerator = prior * p_e_given_h;
  return numerator /
         (numerator + ((1.0 - prior) * p_e_given_not_h));
}

template <typename Operation>
[[nodiscard]] BenchmarkResult benchmark(std::string_view name,
                                        std::size_t iterations,
                                        Operation operation) {
  constexpr std::size_t trial_count = 7;
  const std::size_t warmup_iterations =
      std::min<std::size_t>(iterations, 10'000);
  double checksum = 0.0;
  for (std::size_t index = 0; index < warmup_iterations; ++index) {
    checksum += operation();
  }

  std::vector<double> samples;
  samples.reserve(trial_count);
  for (std::size_t trial = 0; trial < trial_count; ++trial) {
    const auto started = Clock::now();
    for (std::size_t index = 0; index < iterations; ++index) {
      checksum += operation();
    }
    const auto elapsed = Clock::now() - started;
    const double elapsed_nanoseconds =
        std::chrono::duration<double, std::nano>{elapsed}.count();
    samples.push_back(elapsed_nanoseconds / static_cast<double>(iterations));
  }
  std::sort(samples.begin(), samples.end());

  // Printing a checksum after timing keeps the optimizer from discarding work.
  const double posterior = checksum /
                           static_cast<double>(warmup_iterations +
                                               (trial_count * iterations));
  return BenchmarkResult{name, samples[trial_count / 2], posterior};
}

[[nodiscard]] std::size_t parse_iterations(int argument_count,
                                           char* arguments[]) {
  constexpr std::size_t default_iterations = 200'000;
  if (argument_count == 1) {
    return default_iterations;
  }
  if (argument_count != 2) {
    throw std::invalid_argument(
        "usage: belief_update_demo [positive-iteration-count]");
  }

  std::size_t parsed_characters = 0;
  const unsigned long long parsed =
      std::stoull(arguments[1], &parsed_characters, 10);
  if (parsed_characters != std::string{arguments[1]}.size() || parsed == 0) {
    throw std::invalid_argument("iteration count must be a positive integer");
  }
  return static_cast<std::size_t>(parsed);
}

void print_benchmark(const BenchmarkResult& result,
                     std::size_t iterations) {
  std::cout << std::left << std::setw(31) << result.name << std::right
            << std::setw(12) << std::fixed << std::setprecision(2)
            << result.nanoseconds_per_update << " ns/update\n";
  std::cout << "BENCHMARK," << result.name << ',' << std::setprecision(17)
            << result.nanoseconds_per_update << ',' << iterations << ','
            << result.posterior << '\n';
}

}  // namespace

int main(int argument_count, char* arguments[]) {
  using namespace belief_update;

  try {
    const std::size_t iterations =
        parse_iterations(argument_count, arguments);
    const LikelihoodInput explicit_input =
        make_input(0.8, 0.3, SourceType::human, "analyst-7",
                   OverrideStatus::requested);
    const LikelihoodInput model_input =
        make_input(0.8, 0.2, SourceType::model, "model-a",
                   OverrideStatus::not_requested);
    const UpdateConfig configuration{std::nullopt};
    const UpdateResult result = update(
        UpdateInputs{BeliefState{0.4}, explicit_input, model_input,
                     ProvenanceTimestamp{1'000}},
        configuration, UpdateAlgorithm::exact_binary);

    const BayesCalculation& calculation = result.intermediate_calculations();
    const UpdateResult reproduced = replay(result);
    const LocalSensitivityResult sensitivity = local_sensitivity(result);
    const PosteriorComparisonResult comparison =
        compare_model_and_explicit(result);
    const ThresholdAnalysisResult threshold = analyze_threshold(result, 0.6);
    const StressResult stress = stress_update(
        result, RelativeStressRanges{10.0, 10.0, 10.0}, 0.7);
    const BayesCalculation log_odds =
        log_odds_update(BeliefState{0.4}, std::log(0.8 / 0.2));
    const HypothesisTree tree{
        {std::nullopt, HypothesisId{0}, HypothesisId{0}, HypothesisId{2},
         HypothesisId{2}}};
    const TreeBeliefState tree_prior{tree, {0.5, 0.2, 0.3}, 1e-12};
    const TreeLikelihoodValues tree_likelihoods{{0.1, 0.8, 0.4}};
    const Provenance tree_provenance = make_tree_provenance();
    const TreeUpdateResult tree_result =
        update_tree(tree_prior, tree_likelihoods, tree_provenance);
    const TreeUpdateResult replayed_tree = replay(tree_result);

    std::cout << std::setprecision(9);
    std::cout << "BBUS C++ capability demo\n"
              << "  prior:                    " << calculation.prior() << '\n'
              << "  likelihood ratio:         "
              << calculation.likelihood_ratio() << '\n'
              << "  posterior:                " << result.posterior() << '\n'
              << "  belief delta:             " << calculation.belief_delta()
              << '\n'
              << "  provenance source id:     "
              << result.provenance().source_id().value() << '\n'
              << "  replay identical:         " << std::boolalpha
              << (reproduced.posterior() == result.posterior()) << '\n'
              << "  log-odds posterior:       " << log_odds.posterior() << '\n'
              << "  d(posterior)/d(prior):    "
              << sensitivity.d_posterior_d_prior() << '\n'
              << "  model posterior:          "
              << comparison.model_posterior() << '\n'
              << "  explicit posterior:       "
              << comparison.explicit_posterior() << '\n'
              << "  explicit - model:         " << comparison.difference()
              << '\n'
              << "  threshold 0.6 met:        " << threshold.threshold_met()
              << '\n'
              << "  stress minimum posterior: "
              << stress.minimum_posterior_scenario().posterior() << '\n'
              << "  stress maximum posterior: "
              << stress.maximum_posterior_scenario().posterior() << '\n'
              << "  original unchanged:       "
              << (result.posterior() == calculation.posterior()) << '\n'
              << "  tree car posterior:       "
              << tree_result.posterior().probability(HypothesisId{1}) << '\n'
              << "  tree public transit:      "
              << tree_result.posterior().probability(HypothesisId{2}) << '\n'
              << "  tree bus posterior:       "
              << tree_result.posterior().probability(HypothesisId{3}) << '\n'
              << "  tree train posterior:     "
              << tree_result.posterior().probability(HypothesisId{4}) << '\n'
              << "  tree replay identical:    "
              << std::equal(
                     replayed_tree.posterior().node_probabilities().begin(),
                     replayed_tree.posterior().node_probabilities().end(),
                     tree_result.posterior().node_probabilities().begin(),
                     tree_result.posterior().node_probabilities().end())
              << "\n\n";

    volatile double benchmark_prior = 0.4;
    const BeliefState core_prior{0.4};
    const LikelihoodValues core_likelihoods{0.8, 0.2};
    const BenchmarkResult formula_result = benchmark(
        "cpp_unchecked_formula", iterations, [&benchmark_prior] {
          return unchecked_formula(benchmark_prior, 0.8, 0.2);
        });
    const BenchmarkResult core_result = benchmark(
        "bbus_cpp_core", iterations, [&core_prior, &core_likelihoods] {
          return exact_binary_update(core_prior, core_likelihoods).posterior();
        });
    const BenchmarkResult pipeline_result = benchmark(
        "bbus_cpp_audited_update", iterations, [&benchmark_prior] {
          return update(benchmark_prior, 0.8, 0.2).posterior();
        });
    const BenchmarkResult tree_result_benchmark = benchmark(
        "bbus_cpp_tree_update", iterations,
        [&tree_prior, &tree_likelihoods, &tree_provenance] {
          return update_tree(tree_prior, tree_likelihoods, tree_provenance)
              .posterior()
              .probability(HypothesisId{2});
        });

    std::cout << "Median of 7 trials (" << iterations
              << " updates per trial; lower is better)\n";
    print_benchmark(formula_result, iterations);
    print_benchmark(core_result, iterations);
    print_benchmark(pipeline_result, iterations);
    print_benchmark(tree_result_benchmark, iterations);
    std::cout << "\nThe unchecked formula is a latency floor, not a safe or "
                 "auditable replacement.\n";
  } catch (const std::exception& exception) {
    std::cerr << "demo error: " << exception.what() << '\n';
    return EXIT_FAILURE;
  }

  return EXIT_SUCCESS;
}

File: demo/belief_update_demo.py
#!/usr/bin/env python3
"""Demonstrate BBUS and generate reproducible latency comparison artifacts."""

from __future__ import annotations

import argparse
import csv
import html
import math
from pathlib import Path
import platform
import statistics
import subprocess
import sys
import time
from typing import Callable, NamedTuple

import belief_update


class Benchmark(NamedTuple):
    name: str
    median_ns: float
    iterations: int
    posterior: float
    scope: str


def make_input(
    p_e_given_h: float,
    p_e_given_not_h: float,
    source_type: belief_update.SourceType,
    source_id: str,
    override_status: belief_update.OverrideStatus,
) -> belief_update.LikelihoodInput:
    return belief_update.LikelihoodInput(
        likelihoods=belief_update.LikelihoodValues(
            evidence_given_hypothesis=p_e_given_h,
            evidence_given_not_hypothesis=p_e_given_not_h,
        ),
        evidence_weight=None,
        provenance=belief_update.Provenance(
            source_type=source_type,
            source_id=belief_update.SourceId(value=source_id),
            version=belief_update.SourceVersion(value="demo-v1"),
            timestamp=belief_update.ProvenanceTimestamp(
                unix_nanoseconds=1_000
            ),
            confidence=0.95,
            override_status=override_status,
        ),
    )


def showcase() -> None:
    explicit_input = make_input(
        0.8,
        0.3,
        belief_update.SourceType.human,
        "analyst-7",
        belief_update.OverrideStatus.requested,
    )
    model_input = make_input(
        0.8,
        0.2,
        belief_update.SourceType.model,
        "model-a",
        belief_update.OverrideStatus.not_requested,
    )
    result = belief_update.update(
        inputs=belief_update.UpdateInputs(
            prior=belief_update.BeliefState(probability=0.4),
            explicit_override=explicit_input,
            likelihood_model_value=model_input,
            evaluation_time=belief_update.ProvenanceTimestamp(
                unix_nanoseconds=1_000
            ),
        ),
        configuration=belief_update.UpdateConfig(freshness_rule=None),
        algorithm=belief_update.UpdateAlgorithm.exact_binary,
    )
    calculation = result.intermediate_calculations
    replayed = belief_update.replay(result)
    sensitivity = belief_update.local_sensitivity(result=result)
    comparison = belief_update.compare_model_and_explicit(result=result)
    threshold = belief_update.analyze_threshold(result=result, threshold=0.6)
    stress = belief_update.stress_update(
        result=result,
        ranges=belief_update.RelativeStressRanges(
            prior_percent=10.0,
            p_e_given_h_percent=10.0,
            p_e_given_not_h_percent=10.0,
        ),
        threshold=0.7,
    )
    log_odds = belief_update.log_odds_update(
        prior=belief_update.BeliefState(probability=0.4),
        log_likelihood_ratio=math.log(0.8 / 0.2),
    )

    print("BBUS Python capability demo")
    print(f"  prior:                    {calculation.prior:.9f}")
    print(f"  likelihood ratio:         {calculation.likelihood_ratio:.9f}")
    print(f"  posterior:                {result.posterior:.9f}")
    print(f"  belief delta:             {calculation.belief_delta:.9f}")
    print(f"  provenance source id:     {result.provenance.source_id.value}")
    print(f"  replay identical:         {replayed.posterior == result.posterior}")
    print(f"  log-odds posterior:       {log_odds.posterior:.9f}")
    print(f"  d(posterior)/d(prior):    {sensitivity.d_posterior_d_prior:.9f}")
    print(f"  model posterior:          {comparison.model_posterior:.9f}")
    print(f"  explicit posterior:       {comparison.explicit_posterior:.9f}")
    print(f"  explicit - model:         {comparison.difference:.9f}")
    print(f"  threshold 0.6 met:        {threshold.threshold_met}")
    print(
        "  stress minimum posterior: "
        f"{stress.minimum_posterior_scenario.posterior:.9f}"
    )
    print(
        "  stress maximum posterior: "
        f"{stress.maximum_posterior_scenario.posterior:.9f}"
    )
    print(f"  original unchanged:       {result.posterior == calculation.posterior}")
    print()


def time_operation(
    name: str,
    operation: Callable[[], float],
    iterations: int,
    scope: str,
) -> Benchmark:
    warmup_iterations = min(iterations, 1_000)
    checksum = 0.0
    for _ in range(warmup_iterations):
        checksum += operation()

    samples: list[float] = []
    for _ in range(7):
        started = time.perf_counter_ns()
        for _ in range(iterations):
            checksum += operation()
        elapsed = time.perf_counter_ns() - started
        samples.append(elapsed / iterations)

    posterior = checksum / (warmup_iterations + (7 * iterations))
    return Benchmark(name, statistics.median(samples), iterations, posterior, scope)


def unchecked_python_formula() -> float:
    prior = 0.4
    p_e_given_h = 0.8
    p_e_given_not_h = 0.2
    numerator = prior * p_e_given_h
    return numerator / (numerator + ((1.0 - prior) * p_e_given_not_h))


def sklearn_operation() -> tuple[Callable[[], float], str] | None:
    try:
        import numpy
        import sklearn
        from sklearn.naive_bayes import BernoulliNB
    except ImportError:
        return None

    # Integer counts encode prior=0.4, P(E|H)=0.8, and P(E|not H)=0.2.
    observations = numpy.array(
        ([[1.0]] * 12)
        + ([[0.0]] * 48)
        + ([[1.0]] * 32)
        + ([[0.0]] * 8)
    )
    classes = numpy.array(([0] * 60) + ([1] * 40))
    try:
        model = BernoulliNB(alpha=0.0, force_alpha=True)
    except TypeError:
        model = BernoulliNB(alpha=0.0)
    model.fit(observations, classes)
    sample = numpy.array([[1.0]])

    def infer() -> float:
        return float(model.predict_proba(sample)[0, 1])

    expected = 8.0 / 11.0
    if not math.isclose(infer(), expected, rel_tol=1e-9, abs_tol=1e-12):
        raise RuntimeError("scikit-learn comparison model has the wrong posterior")
    return infer, sklearn.__version__


def run_cpp_demo(executable: Path, iterations: int) -> list[Benchmark]:
    completed = subprocess.run(
        [str(executable), str(iterations)],
        check=True,
        capture_output=True,
        text=True,
    )
    print(completed.stdout.rstrip())
    results: list[Benchmark] = []
    scopes = {
        "cpp_unchecked_formula": "unchecked arithmetic latency floor",
        "bbus_cpp_core": "validated C++ mathematical core",
        "bbus_cpp_audited_update": "complete C++ validate-resolve-update-result path",
    }
    for line in completed.stdout.splitlines():
        if not line.startswith("BENCHMARK,"):
            continue
        _, name, nanoseconds, count, posterior = line.split(",")
        results.append(
            Benchmark(
                name,
                float(nanoseconds),
                int(count),
                float(posterior),
                scopes[name],
            )
        )
    if len(results) != 3:
        raise RuntimeError("C++ demo did not emit all benchmark records")
    return results


def format_duration(nanoseconds: float) -> str:
    if nanoseconds < 1_000.0:
        return f"{nanoseconds:.1f} ns"
    if nanoseconds < 1_000_000.0:
        return f"{nanoseconds / 1_000.0:.2f} us"
    return f"{nanoseconds / 1_000_000.0:.2f} ms"


def write_csv(results: list[Benchmark], destination: Path) -> None:
    with destination.open("w", newline="", encoding="utf-8") as output:
        writer = csv.writer(output, lineterminator="\n")
        writer.writerow(
            ["implementation", "median_ns_per_update", "iterations", "posterior", "scope"]
        )
        writer.writerows(results)


def write_markdown(
    results: list[Benchmark], destination: Path, sklearn_version: str | None
) -> None:
    lines = [
        "# Demo benchmark results",
        "",
        "Median of seven trials. Lower latency is better.",
        "",
        "| Implementation | Median per update | Iterations/trial | Posterior | Scope |",
        "|---|---:|---:|---:|---|",
    ]
    for result in sorted(results, key=lambda item: item.median_ns):
        lines.append(
            f"| `{result.name}` | {format_duration(result.median_ns)} | "
            f"{result.iterations:,} | {result.posterior:.9f} | {result.scope} |"
        )
    lines.extend(
        [
            "",
            f"- Platform: `{platform.platform()}`",
            f"- Python: `{platform.python_version()}`",
            f"- BBUS: `{belief_update.__version__}`",
            f"- scikit-learn: `{sklearn_version or 'not installed; comparison skipped'}`",
            "",
            "These are illustrative local measurements, not universal performance claims. "
            "The scikit-learn row measures scalar `BernoulliNB.predict_proba`; it does not "
            "provide BBUS provenance, validation, replay, or stress-analysis semantics. "
            "Unchecked formula rows omit safety and audit behavior.",
            "",
        ]
    )
    destination.write_text("\n".join(lines), encoding="utf-8")


def write_svg(results: list[Benchmark], destination: Path) -> None:
    ordered = sorted(results, key=lambda item: item.median_ns)
    width = 1_080
    left = 295
    right = 175
    top = 105
    row_height = 58
    bottom = 90
    chart_width = width - left - right
    height = top + (row_height * len(ordered)) + bottom
    minimum_power = math.floor(math.log10(min(item.median_ns for item in ordered)))
    maximum_power = math.ceil(math.log10(max(item.median_ns for item in ordered)))
    if minimum_power == maximum_power:
        maximum_power += 1

    def x_position(value: float) -> float:
        fraction = (math.log10(value) - minimum_power) / (
            maximum_power - minimum_power
        )
        return left + (fraction * chart_width)

    colors = {
        "bbus_cpp_core": "#2f6fed",
        "bbus_cpp_audited_update": "#1854b4",
        "bbus_python_core_binding": "#23a98c",
        "bbus_python_audited_update": "#147a65",
        "sklearn_bernoulli_nb_scalar": "#e28a24",
    }
    svg = [
        f'<svg xmlns="http://www.w3.org/2000/svg" width="{width}" height="{height}" viewBox="0 0 {width} {height}">',
        '<rect width="100%" height="100%" fill="#fbfaf7"/>',
        '<text x="40" y="42" font-family="system-ui, sans-serif" font-size="24" font-weight="700" fill="#172033">BBUS demo latency comparison</text>',
        '<text x="40" y="70" font-family="system-ui, sans-serif" font-size="14" fill="#536078">Median ns per single update - lower is better - logarithmic scale</text>',
    ]
    for power in range(minimum_power, maximum_power + 1):
        value = 10.0**power
        x = x_position(value)
        svg.append(
            f'<line x1="{x:.2f}" y1="{top - 15}" x2="{x:.2f}" y2="{height - bottom + 10}" stroke="#d9dde5" stroke-width="1"/>'
        )
        svg.append(
            f'<text x="{x:.2f}" y="{height - 42}" text-anchor="middle" font-family="ui-monospace, monospace" font-size="12" fill="#667085">{html.escape(format_duration(value))}</text>'
        )

    for index, result in enumerate(ordered):
        y = top + (index * row_height)
        end_x = x_position(result.median_ns)
        color = colors.get(result.name, "#7a8394")
        svg.append(
            f'<text x="{left - 14}" y="{y + 23}" text-anchor="end" font-family="ui-monospace, monospace" font-size="13" fill="#263247">{html.escape(result.name)}</text>'
        )
        svg.append(
            f'<rect x="{left}" y="{y + 5}" width="{max(3.0, end_x - left):.2f}" height="26" rx="4" fill="{color}"/>'
        )
        svg.append(
            f'<text x="{min(width - right + 12, end_x + 10):.2f}" y="{y + 23}" font-family="ui-monospace, monospace" font-size="13" font-weight="700" fill="#172033">{html.escape(format_duration(result.median_ns))}</text>'
        )
    svg.extend(
        [
            f'<text x="40" y="{height - 12}" font-family="system-ui, sans-serif" font-size="12" fill="#667085">Local illustrative benchmark; scopes differ. See benchmark_results.md.</text>',
            "</svg>",
        ]
    )
    destination.write_text("\n".join(svg), encoding="utf-8")


def parse_arguments() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--iterations", type=int, default=10_000)
    parser.add_argument("--cpp-iterations", type=int, default=200_000)
    parser.add_argument("--cpp-executable", type=Path)
    parser.add_argument(
        "--output-dir", type=Path, default=Path(__file__).parent / "results"
    )
    parser.add_argument("--skip-sklearn", action="store_true")
    arguments = parser.parse_args()
    if arguments.iterations <= 0 or arguments.cpp_iterations <= 0:
        parser.error("iteration counts must be positive")
    return arguments


def main() -> int:
    arguments = parse_arguments()
    showcase()

    prior = belief_update.BeliefState(probability=0.4)
    likelihoods = belief_update.LikelihoodValues(
        evidence_given_hypothesis=0.8,
        evidence_given_not_hypothesis=0.2,
    )
    results = [
        time_operation(
            "python_unchecked_formula",
            unchecked_python_formula,
            arguments.iterations,
            "unchecked Python arithmetic latency floor",
        ),
        time_operation(
            "bbus_python_core_binding",
            lambda: belief_update.exact_binary_update(
                prior=prior, likelihoods=likelihoods
            ).posterior,
            arguments.iterations,
            "Python-to-C++ mathematical core binding",
        ),
        time_operation(
            "bbus_python_audited_update",
            lambda: belief_update.update(
                prior=0.4, p_e_given_h=0.8, p_e_given_not_h=0.2
            ).posterior,
            arguments.iterations,
            "Python-to-C++ complete validate-resolve-update-result path",
        ),
    ]

    sklearn_version: str | None = None
    if not arguments.skip_sklearn:
        comparison = sklearn_operation()
        if comparison is None:
            print(
                "scikit-learn is not installed; its optional comparison was skipped."
            )
        else:
            operation, sklearn_version = comparison
            results.append(
                time_operation(
                    "sklearn_bernoulli_nb_scalar",
                    operation,
                    arguments.iterations,
                    "scalar BernoulliNB.predict_proba on equivalent evidence",
                )
            )

    if arguments.cpp_executable is not None:
        results.extend(
            run_cpp_demo(arguments.cpp_executable, arguments.cpp_iterations)
        )

    print("\nMedian of 7 trials (lower is better)")
    for result in sorted(results, key=lambda item: item.median_ns):
        print(f"  {result.name:<31} {format_duration(result.median_ns):>10}")

    arguments.output_dir.mkdir(parents=True, exist_ok=True)
    csv_path = arguments.output_dir / "benchmark_results.csv"
    markdown_path = arguments.output_dir / "benchmark_results.md"
    svg_path = arguments.output_dir / "benchmark_results.svg"
    write_csv(results, csv_path)
    write_markdown(results, markdown_path, sklearn_version)
    write_svg(results, svg_path)
    print(f"\nWrote {csv_path}, {markdown_path}, and {svg_path}")
    return 0


if __name__ == "__main__":
    sys.exit(main())
